File: examples/go/language_gauntlet/internal/storage/storage.go
package storage

import (
	domainmodel "example.com/bonsai/language-gauntlet/internal/domain"
	execflow "example.com/bonsai/language-gauntlet/internal/runtime"
)

type Runner interface {
	Run() int
}

var _ Runner = (*AuditedRepository)(nil)

type Repository struct {
	data domainmodel.Envelope
}

func NewRepository(data domainmodel.Envelope) *Repository {
	return &Repository{data: data}
}

func (repository *Repository) Cmd() string {
	return repository.data.Command()
}

func (repository *Repository) Run() int {
	cmd := repository.Cmd()
	return execflow.Execute(cmd)
}

type AuditedRepository struct {
	*Repository
}

func NewAuditedRepository(data domainmodel.Envelope) *AuditedRepository {
	return &AuditedRepository{Repository: NewRepository(data)}
}

func (repository *AuditedRepository) Run() int {
	return repository.Repository.Run()
}

func Persist(data domainmodel.Envelope) int {
	repository := NewAuditedRepository(data)
	return repository.Run()
}

File: examples/c/language_gauntlet/src/storage/storage.c
#include "gauntlet/envelope.h"
#include "gauntlet/executor.h"
#include "gauntlet/storage.h"

struct Repository {
    EnvelopeAlias data;
    const char *command;
    int (*run)(const struct Repository *self, const char *command);
};

static int repository_run(const struct Repository *self, const char *command) {
    const char *cmd = self->command != NULL ? command : self->data.selected;
    return execute(cmd);
}

static struct Repository repository_create(const EnvelopeAlias *env, const char *command) {
    struct Repository repo = {*env, command, repository_run};
    return repo;
}

int persist(const EnvelopeAlias *env, const char *command) {
    struct Repository repo = repository_create(env, command);
    int (*runner)(const struct Repository *self, const char *command) = repo.run;
    (void)runner;
    return repository_run(&repo, command);
}
